pisek/cms/__init__.py

```diff
@@ -1,4 +1,181 @@
+#
+# Copyright (c)   2024        Benjamin Swart <Benjaminswart@email.cz>
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# any later version.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+# The CMS module imports and calls internal CMS libraries,
+# as well as dependencies of CMS (sqlalchemy).
+# Since these dependencies aren't declared anywhere, importing this module
+# will cause an exception on most installations.
+# It is therefore important not to import this module unless a cms command has actually been called.
+
+# The implementation uses CMS's sqlalchemy types to make changes to the database.
+# It also uses some other APIs, like session creation.
+# If CMS changes its database schema or some of its internal APIs,
+# this module will likely break.
+
+# This module obviously also uses internal methods of pisek,
+# mostly config parsing and some general utilities.
+# Large refactors in pisek will also require changes to this module.
+
+# Some functionality is duplicated, most notably enumerating all inputs
+# and finding the file extensions of solutions.
+# All of this functionality is implemented inside private methods of pipeline jobs,
+# making it inaccessible to part of Pisek that don't use the pipeline.
+
+# The pipeline is called when creating a dataset, in order to generate inputs and outputs
+# and to compile the judge.
+# It is called as if the user requested the primary solution to be tested.
+
+# One small thing to note is that which automatic submission corresponds
+# to which solution isn't explicitly stored anywhere.
+# Instead, the check and testing-log commands simply search the database for
+# a matching (task, file hash, language) triple.
+
+from typing import Any, Callable
+from cms.db.session import Session
+from sqlalchemy.exc import IntegrityError
+
+from pisek.cms.dataset import create_dataset, get_dataset
+from pisek.cms.result import create_testing_log, check_results
+from pisek.cms.submission import get_participation, submit_all
+from pisek.cms.task import create_task, get_task, set_task_settings
+from pisek.env.env import Env
+from pisek.jobs.cache import Cache
+from pisek.jobs.task_pipeline import TaskPipeline
+from pisek.utils.pipeline_tools import PATH
+
+
+def prepare_files(env: Env):
+    contest_type = env.config.contest_type
+
+    if contest_type != "cms":
+        raise RuntimeError(f"Cannot upload {contest_type}-type task to CMS")
+
+    env = env.fork()
+    env.solutions = [env.config.primary_solution]
+
+    if TaskPipeline(env).run_jobs(Cache(env), env) != 0:
+        raise RuntimeError("Failed to test primary solution, cannot upload to CMS")
+
+
+def with_env(fun: Callable[[Env, Any], int]) -> Callable[[Any], int]:
```

Comment:
This decorator is only really usefull for commands, which have a pretty specific signature.
The decorator requires that the argument is an `argparse.Namespace` or something dictionary-like
and that the return type is an `int`.

Reply:
My bad, right.

Reply:
We could do something like `Param = ParamSpec("Param", Namespace)` (possibly with a protocol for classes that support `__dict__` instead of `Namespace`) in case we ever want to add better type hints for arguments in the future.
We could also add support for other arguments.